src/bsoncxx/lib/bsoncxx/v1/types/value.cpp

```diff
@@ -12,19 +12,470 @@
+            other._value = {BSON_TYPE_NULL, {}, {}};
+        }
+        return *this;
+    }
+
+    impl(impl const& other) {
+        bson_value_copy(&other._value, &_value);
+    }
+
+    impl& operator=(impl const& other) {
+        if (&other != this) {
+            bson_value_destroy(&_value);
+            bson_value_copy(&other._value, &_value);
+        }
+        return *this;
+    }
+
+    impl() = default;
+
+    auto t() const -> decltype((_value.value_type)) {
+        return _value.value_type;
+    }
+
+    auto t() -> decltype((_value.value_type)) {
+        return _value.value_type;
+    }
+
+    auto v() const -> decltype((_value.value)) {
+        return _value.value;
+    }
+
+    auto v() -> decltype((_value.value)) {
+        return _value.value;
+    }
+};
+
+value::impl::~impl() {
+    static_assert(is_semiregular<impl>::value, "bsoncxx::v1::types::value::impl must be semiregular");
+    static_assert(is_nothrow_moveable<impl>::value, "bsoncxx::v1::types::value::impl must be nothrow moveable");
+
+    static_assert(sizeof(value::_storage) >= sizeof(value::impl), "insufficient size");
+    static_assert(alignof(value) >= alignof(value::impl), "insufficient alignment");
+
+    bson_value_destroy(&_value);
+}
+
+value::~value() {
+    internal::impl(this)->~impl();
+}
+
+value::value(value&& other) noexcept {
+    new (internal::impl(this)) impl{std::move(internal::impl(other))};
+}
+
+value& value::operator=(value&& other) noexcept {
+    *internal::impl(this) = std::move(internal::impl(other));
+    return *this;
+}
+
+value::value(value const& other) {
+    new (internal::impl(this)) impl{internal::impl(other)};
+}
+
+value& value::operator=(value const& other) {
+    *internal::impl(this) = internal::impl(other);
+    return *this;
+}
+
+namespace {
+
+// For backward compatibility, do not prematurely truncate strings in bsoncxx API. Instead, defer handling of potential
+// embedded null bytes to the bson library.
+char* to_bson_copy(v1::stdx::string_view sv) {
+    if (sv.empty()) {
+        return nullptr;
+    }
+
+    auto ret = static_cast<char*>(bson_malloc(sv.size() + 1u));
+    std::memcpy(ret, sv.data(), sv.size());
+    ret[sv.size()] = '\0';
+    return ret;
+}
+
+std::uint8_t* to_bson_copy(void const* data, std::size_t size) {
+    if (!data) {
+        return nullptr;
+    }
+
+    auto ret = static_cast<std::uint8_t*>(bson_malloc(size));
+    std::memcpy(ret, data, size);
+    return ret;
+}
+
+} // namespace
+
+value::value(v1::types::view const& v) : value{} {
+#pragma push_macro("X")
+#undef X
+#define X(_name, _value)                \
+    case v1::types::id::k_##_name:      \
+        *this = value{v.get_##_name()}; \
+        return;
+
+    switch (v.type_id()) {
+        BSONCXX_V1_TYPES_XMACRO(X)
+
+        default:
+            throw v1::exception{code::invalid_type};
+    }
+#pragma pop_macro("X")
+}
+
+// BSONCXX_V1_TYPES_XMACRO: update below.
```

Comment:
The overloads below implement the behavior currently handled by v_noabi's `convert_to_libbson()` in [convert.hh](https://github.com/mongodb/mongo-cxx-driver/blob/ad622035ec9c9ab352b42449d5fcbf426e8782e9/src/bsoncxx/lib/bsoncxx/private/convert.hh), but without exposing `bson_value_t` to dependent components.